### PR TITLE
Enhance `kytos_stats`

### LINKED ISSUE #243: Enhance e2e tests for `kytos_stats`
This is to add a test on e2e test suit for `kytos_stats`, according to issue #242.

First, information will be collected from `kytos_stats`. Then, some traffic will be created between hosts, and collected information again from `kytos_stats`. This will allow us to make sure that the stats are accurate. 
It will also add a test in which a flow is added to later query it for stats, instead of trying to find an existing flow as current.

Plus: add some text for failure on assert, to help troubleshooting later.

### PR DESCRIPTION
**Pull request by gretelliz**
Closes #243 

### Summary

Add `test_036_table_1_active_count_update` test to create some traffic on table 1, and  make sure that the stats are accurate.

Update `test_015_packet_count` and `test_020_bytes_count` tests to avoid trying to find a non-empty flow, instead a flow is installed and query it later for stats.

### CODE REVIEW

File: tests/test_e2e_70_kytos_stats.py [hunk 1]
```diff
@@ -62,16 +62,16 @@ def test_010_table_stats(self):
         response = requests.get(api_url)
         assert response.status_code == 200, response.text
         data = response.json()
-        assert len(data) == 1
-        assert '00:00:00:00:00:00:00:01' in data
+        assert len(data) == 1, str(data)
+        assert '00:00:00:00:00:00:00:01' in data, str(data)
```
Comment (viniarck) [inline tests/test_e2e_70_kytos_stats.py:66]: Typically we only add extra context to the assert messages when pytest doesn't add much info given a boolean comparison like `len(data) == 1`. For instance, in a string `in` assertion it typically includes, even in older version of pytest:

```
E       AssertionError: assert 'a' in 'some'
```

The extra content won't cause harm though, but if we don't have too then we're just adding unnecessary information, can you simulate an assertion error for this equivalent case and see if it's really needed?

File: tests/test_e2e_70_kytos_stats.py [hunk 2]
```diff
@@ -82,78 +82,104 @@ def test_010_table_stats(self):
...
+        assert len(data) == len(topo_switches), str(data)
 
     def test_015_packet_count(self):
         """Test packet_count""" 
+        sw = "00:00:00:00:00:00:00:01"
 
-        api_url = KYTOS_STATS + '/flow/stats'  
+        # install a flow
+        cookie = 5
+        payload = {"flows": [{"cookie": cookie, "packet_count": 8, "duration_sec": 2}]}
```
Comment (viniarck) [inline tests/test_e2e_70_kytos_stats.py:120]: packet_count and duration_sec shouldn't be sent, it doesn't cause any issues (now), but can get confusing for anyone else reading this in the future, can you review this?

**Review by viniarck — COMMENTED** [review]
The augmented coverage looks good to me. I'll leave it up to @italovalcy to approve though since you both have been discussing about this. Other than that, I raised a few minor points. Great PR, Gretel.

**Comment by gretelliz** [comment]
@viniarck , Thank you.

**Comment by viniarck** [comment]
@gretelliz check out with @italovalcy if this can already land, thanks.

File: tests/test_e2e_70_kytos_stats.py [hunk 3]
```diff
@@ -113,16 +113,29 @@ def test_010_table_stats(self):
 
     def test_015_packet_count(self):
         """Test packet_count""" 
+        sw = "00:00:00:00:00:00:00:01"
 
-        api_url = KYTOS_STATS + '/flow/stats'  
+        # install a flow
+        cookie = 5
+        payload = {"flows": [{"cookie": cookie}]}
```
Comment (italovalcy) [inline tests/test_e2e_70_kytos_stats.py:120]:
```suggestion
        payload = {
            "flows": [{
                "cookie": cookie,
                "match": {"in_port": 1, 'dl_dst': '33:33:00:00:00:02', 'dl_type': 0x86dd},
                'actions': [{'action_type': 'output', 'port': 2}]
            }]
        }
```

File: tests/test_e2e_70_kytos_stats.py [hunk 4]
```diff
@@ -113,16 +113,29 @@ def test_010_table_stats(self):
...
+
+        api_url_flow_manager = KYTOS_API + f'/kytos/flow_manager/v2/flows/{sw}'
+        response = requests.post(api_url_flow_manager, data=json.dumps(payload),
+                                 headers={'Content-type': 'application/json'})
+        assert response.status_code == 202, response.text
+        data_flow = response.json()
+        assert 'FlowMod Messages Sent' in data_flow['response']
+
+        time.sleep(10)
```
Comment (italovalcy) [inline tests/test_e2e_70_kytos_stats.py:129]:
```suggestion
        # wait the flow to be installed
        time.sleep(5)

        # send N packets, each one containing 1500 bytes
        # (14 ether hdr + 40 ipv6 + 8 icmp + 1438 payload)
        h11 = self.net.net.get('h11')
        n = 20
        h11.cmd(f"ping -6 -b -c {n} -s 1438 FF02::2%h11-eth0 -Mdo -i 0.01 -W 2")

        # give enough time for stats gathering (of_core.STATS_INTERVAL)
        time.sleep(10)
```

File: tests/test_e2e_70_kytos_stats.py [hunk 5]
```diff
@@ -135,16 +148,29 @@ def test_015_packet_count(self):
 
     def test_020_bytes_count(self):
         """Test bytes_count""" 
+        sw = "00:00:00:00:00:00:00:01"
 
-        api_url = KYTOS_STATS + '/flow/stats'  
+        # install a flow
+        cookie = 5
+        payload = {"flows": [{"cookie": cookie}]}
```
Comment (italovalcy) [inline tests/test_e2e_70_kytos_stats.py:155]:
```suggestion
        payload = {
            "flows": [{
                "cookie": cookie,
                "match": {"in_port": 1, 'dl_dst': '33:33:00:00:00:02', 'dl_type': 0x86dd},
                'actions': [{'action_type': 'output', 'port': 2}]
            }]
        }
```

File: tests/test_e2e_70_kytos_stats.py [hunk 6]
```diff
@@ -135,16 +148,29 @@ def test_015_packet_count(self):
...
+
+        api_url_flow_manager = KYTOS_API + f'/kytos/flow_manager/v2/flows/{sw}'
+        response = requests.post(api_url_flow_manager, data=json.dumps(payload),
+                                 headers={'Content-type': 'application/json'})
+        assert response.status_code == 202, response.text
+        data_flow = response.json()
+        assert 'FlowMod Messages Sent' in data_flow['response']
+
+        time.sleep(10)
```
Comment (italovalcy) [inline tests/test_e2e_70_kytos_stats.py:164]:
```suggestion
        # wait the flow to be installed
        time.sleep(5)

        # send N packets, each one containing 1500 bytes
        # (14 ether hdr + 40 ipv6 + 8 icmp + 1438 payload)
        h11 = self.net.net.get('h11')
        n = 20
        h11.cmd(f"ping -6 -b -c {n} -s 1438 FF02::2%h11-eth0 -Mdo -i 0.01 -W 2")

        # waiting to give enough time for stats gathering (of_core.STATS_INTERVAL)
        time.sleep(10)
```

File: tests/test_e2e_70_kytos_stats.py [hunk 7]
```diff
@@ -135,16 +148,29 @@ def test_015_packet_count(self):
...
-        data_flow = response.json()
-        for sw in data_flow:
-            if len(data_flow[sw]):
-                flow_id = list(data_flow[sw].keys())[0]
-                bytes_counter = data_flow[sw][flow_id]['byte_count']
-                bits_per_second = 8*bytes_counter/data_flow[sw][flow_id]['duration_sec']
+        data_flow = response.json()[sw]
+        for flow_id, flow in data_flow.items():
+            if flow['cookie'] == cookie:
+                bytes_counter = flow['byte_count']
```
Comment (italovalcy) [inline tests/test_e2e_70_kytos_stats.py:172]:
```suggestion
                bytes_counter = flow['byte_count']
                assert bytes_counter >= n*1500, str(flow)
```

File: tests/test_e2e_70_kytos_stats.py [hunk 8]
```diff
@@ -113,16 +113,29 @@ def test_010_table_stats(self):
...
-        data_flow = response.json()
-        for sw in data_flow:
-            if len(data_flow[sw]):
-                flow_id = list(data_flow[sw].keys())[0]
-                packet_counter = data_flow[sw][flow_id]['packet_count']
-                packet_per_second = packet_counter/data_flow[sw][flow_id]['duration_sec']
+        data_flow = response.json()[sw]
+        for flow_id, flow in data_flow.items():
+            if flow['cookie'] == cookie:
+                packet_counter = flow['packet_count']
```
Comment (italovalcy) [inline tests/test_e2e_70_kytos_stats.py:137]:
```suggestion
                packet_counter = flow['packet_count']
                assert packet_counter >= n, str(flow)
```

**Review by italovalcy — CHANGES_REQUESTED** [review]
Hi @gretelliz thanks for sending the PR! All looks good to me. I've just added some suggestions regarding the inclusion of data-plane tests to also make sure the stats are being collected properly.

**Comment by gretelliz** [comment]
Thank you, @italovalcy. Good suggestion for incorporating dataplane checks.

**Review by viniarck — APPROVED** [review]
@gretelliz looks like this PR is ready to land since you've also addressed @italovalcy's feedback. Please re-run this test suite if it's still passing that feel free to merge.

**Comment by gretelliz** [comment]
> @gretelliz looks like this PR is ready to land since you've also addressed @italovalcy's feedback. Please re-run this test suite if it's still passing that feel free to merge.

```
tests/test_e2e_70_kytos_stats.py ........                                [100%]

=============================== warnings summary ===============================
test_e2e_70_kytos_stats.py: 17 warnings
  /usr/lib/python3/dist-packages/mininet/node.py:1121: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    return ( StrictVersion( cls.OVSVersion ) <

test_e2e_70_kytos_stats.py: 17 warnings
  /usr/lib/python3/dist-packages/mininet/node.py:1122: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    StrictVersion( '1.10' ) )

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------------------------- start/stop times -------------------------------
================== 8 passed, 34 warnings in 467.41s (0:07:47) ==================
```

**Pull request merged**